ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,801 @@
+		registerRequest.ContainerInstanceArn = &containerInstanceArn
+	} else {
+		// This is a new instance, not previously registered.
+		// Custom attribute registration only happens on initial instance registration.
+		for _, attribute := range client.getCustomAttributes() {
+			logger.Debug("Added a new custom attribute", logger.Fields{
+				field.AttributeName:  aws.StringValue(attribute.Name),
+				field.AttributeValue: aws.StringValue(attribute.Value),
+			})
+			registrationAttributes = append(registrationAttributes, attribute)
+		}
+	}
+	// Standard attributes are included with all registrations.
+	registrationAttributes = append(registrationAttributes, attributes...)
+
+	// Add additional attributes, such as the OS type.
+	registrationAttributes = append(registrationAttributes, client.getAdditionalAttributes()...)
+	registrationAttributes = append(registrationAttributes, client.getOutpostAttribute(outpostARN)...)
+
+	registerRequest.Attributes = registrationAttributes
+	if len(tags) > 0 {
+		registerRequest.Tags = tags
+	}
+	registerRequest.PlatformDevices = platformDevices
+	registerRequest = client.setInstanceIdentity(registerRequest)
+
+	resources, err := client.getResources()
+	if err != nil {
+		return "", "", err
+	}
+
+	registerRequest.TotalResources = resources
+
+	registerRequest.ClientToken = &registrationToken
+	resp, err := client.standardClient.RegisterContainerInstance(&registerRequest)
+	if err != nil {
+		logger.Error("Unable to register as a container instance with ECS", logger.Fields{
+			field.Error: err,
+		})
+		return "", "", err
+	}
+
+	var availabilityzone = ""
+	if resp != nil {
+		for _, attr := range resp.ContainerInstance.Attributes {
+			if aws.StringValue(attr.Name) == azAttrName {
+				availabilityzone = aws.StringValue(attr.Value)
+				break
+			}
+		}
+	}
+
+	logger.Info("Registered container instance with cluster!")
+	err = validateRegisteredAttributes(registerRequest.Attributes, resp.ContainerInstance.Attributes)
+	return aws.StringValue(resp.ContainerInstance.ContainerInstanceArn), availabilityzone, err
+}
+
+func (client *ecsClient) setInstanceIdentity(
+	registerRequest ecsmodel.RegisterContainerInstanceInput) ecsmodel.RegisterContainerInstanceInput {
+	instanceIdentityDoc := ""
+	instanceIdentitySignature := ""
+
+	if client.configAccessor.NoInstanceIdentityDocument() {
+		logger.Info("Fetching Instance ID Document has been disabled")
+		registerRequest.InstanceIdentityDocument = &instanceIdentityDoc
+		registerRequest.InstanceIdentityDocumentSignature = &instanceIdentitySignature
+		return registerRequest
+	}
+
+	iidRetrieved := true
+	backoff := retry.NewExponentialBackoff(setInstanceIdRetryBackoffMin, setInstanceIdRetryBackoffMax,
+		setInstanceIdRetryBackoffJitter, setInstanceIdRetryBackoffMultiple)
+	ctx, cancel := context.WithTimeout(context.Background(), setInstanceIdRetryTimeOut)
+	defer cancel()
+	err := retry.RetryWithBackoffCtx(ctx, backoff, func() error {
```

Comment:
This is a new behavior for containerd agent. Was this addressed in the design? I am unsure of the impact of this.

Reply:
This was introduced as part of https://github.com/aws/amazon-ecs-agent/pull/3795 and is universally a good  mechanism to have in place to improve resiliency in the case of intermittent upstream failures. We can chat more about this offline if you need further details than what's listed in that pull request's description.